cpp/src/ray/util/function_helper.cc

```diff
@@ -1,94 +1,153 @@
+    RAY_LOG(WARNING) << "Get execute function failed, lib_path: " << lib_path
+                     << ", unknown failed reason.";
+  }
+  return;
 }
 
-std::string FunctionHelper::OutputAllRemoteFunctionNames(
-    const boost::dll::shared_library &lib) {
-  auto get_names_func =
-      boost::dll::import_alias<std::vector<std::string>()>(lib, "GetRemoteFunctionNames");
+std::string FunctionHelper::LoadAllRemoteFunctions(const std::string lib_path,
+                                                   const boost::dll::shared_library &lib,
+                                                   const EntryFuntion &entry_function) {
+  static const std::string internal_function_name = "GetRemoteFunctions";
+  if (!lib.has(internal_function_name)) {
+    RAY_LOG(WARNING) << "Internal function '" << internal_function_name
+                     << "' not found in " << lib_path;
+    return "";
+  }
+  auto get_remote_func = boost::dll::import_alias<
+      std::pair<const RemoteFunctionMap_t &, const RemoteMemberFunctionMap_t &>()>(
+      lib, internal_function_name);
   std::string names_str;
-  auto names = get_names_func();
-  for (const auto &name : names) {
-    names_str.append(name).append(", ");
+  auto function_maps = get_remote_func();
+  for (const auto &pair : function_maps.first) {
+    names_str.append(pair.first).append(", ");
+    remote_funcs_.emplace(pair.first, entry_function);
+  }
+  for (const auto &pair : function_maps.second) {
+    names_str.append(pair.first).append(", ");
+    remote_member_funcs_.emplace(pair.first, entry_function);
+  }
+  if (!names_str.empty()) {
+    names_str.pop_back();
+    names_str.pop_back();
   }
-  names_str.pop_back();
-  names_str.pop_back();
   return names_str;
 }
 
-std::function<msgpack::sbuffer(const std::string &, const std::vector<msgpack::sbuffer> &,
-                               msgpack::sbuffer *)>
-FunctionHelper::GetEntryFunction(const std::string &lib_name) {
-  auto it = funcs_.find(lib_name);
-  if (it != funcs_.end()) {
-    return it->second;
+void FindDynamicLibrary(boost::filesystem::path path,
+                        std::list<boost::filesystem::path> &dynamic_libraries) {
+#if defined(_WIN32)
+  static const std::unordered_set<std::string> dynamic_library_extension = {".dll"};
+#elif __APPLE__
+  static const std::unordered_set<std::string> dynamic_library_extension = {".dylib",
+                                                                            ".so"};
+#else
+  static const std::unordered_set<std::string> dynamic_library_extension = {".so"};
+#endif
+  auto extension = boost::filesystem::extension(path);
+  if (dynamic_library_extension.find(extension) != dynamic_library_extension.end()) {
+    RAY_LOG(INFO) << path << " dynamic library found.";
+    dynamic_libraries.emplace_back(path);
   }
+}
 
-  auto lib = LoadDll(lib_name);
-  if (lib == nullptr) {
-    return nullptr;
+void FunctionHelper::LoadFunctionsFromPaths(const std::vector<std::string> paths) {
```

Comment:
Why not accept `const std::vector<std::string> &paths` and avoid copying in this function?

Reply:
My fault. Thanks.